### PR TITLE
add System.AddonIsEnabled() infobool and EnableAddon() builtin

### PR DESCRIPTION
**Pull request by ronie**
as requested by @jurialmunkey and other skinners (ref: https://forum.kodi.tv/showthread.php?tid=330706)

- System.AddonIsEnabled() infobool
- EnableAddon() builtin

### CODE REVIEW

**Comment by ksooo** [comment]
We should take care that this survives my guiinfomanager refactoring PR. ;-)

**Comment by ronie** [comment]
sure no worries, will rebase if needed.

File: xbmc/GUIInfoManager.cpp [hunk 1]
```diff
@@ -7758,6 +7764,13 @@ bool CGUIInfoManager::GetMultiInfoBool(const GUIInfo &info, int contextWindow, c
         bReturn = CServiceBroker::GetAddonMgr().GetAddon(m_stringParameters[info.GetData1()],addon) && addon;
         break;
       }
+      case SYSTEM_ADDON_IS_ENABLED:
+      {
+        AddonPtr addon;
+        bReturn = CServiceBroker::GetAddonMgr().GetAddon(m_stringParameters[info.GetData1()],addon) && 
+          !CServiceBroker::GetAddonMgr().IsAddonDisabled(m_stringParameters[info.GetData1()]) && addon;
+        break;
+      }
```
Comment (garbear) [inline xbmc/GUIInfoManager.cpp]: Some refactoring to improve readability (untested):

```c++
const std::string &addonId = m_stringParameters[info.GetData1()];

bReturn = false;

AddonPtr addon;
if (CServiceBroker::GetAddonMgr().GetAddon(addonId, addon))
{
  if (!CServiceBroker::GetAddonMgr().IsAddonDisabled(addonId))
    bReturn = true;
}

break;
```

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 2]
```diff
@@ -64,6 +67,30 @@ static int InstallAddon(const std::vector<std::string>& params)
...
+ */
+static int EnableAddon(const std::vector<std::string>& params)
+{
+  const std::string& addonid = params[0];
+
+  if (!g_passwordManager.CheckMenuLock(WINDOW_ADDON_BROWSER))
+    return -1;
+
+  AddonPtr addon;
+  if (!CServiceBroker::GetAddonMgr().GetAddon(addonid, addon, ADDON_UNKNOWN, false) || addon == nullptr)
```
Comment (garbear) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: You can drop the `addon == nullptr` part, non-null addon is guaranteed if GetAddon() returns true

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 3]
```diff
@@ -64,6 +67,30 @@ static int InstallAddon(const std::vector<std::string>& params)
...
+    return -1;
+
+  if (HELPERS::ShowYesNoDialogLines(CVariant{24076}, CVariant{24135}, CVariant{addon->Name()}, CVariant{24136}) ==
+    DialogResponse::YES)
+  {
+    CServiceBroker::GetAddonMgr().EnableAddon(addonid);
+  }
```
Comment (garbear) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: Another suggestion to improve readability (I can do these all day :) ):

```c++
auto response = HELPERS::ShowYesNoDialogLines(CVariant{24076}, CVariant{24135}, CVariant{addon->Name()}, CVariant{24136};
if (response == DialogResponse::YES)
  CServiceBroker::GetAddonMgr().EnableAddon(addonid);
```

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 4]
```diff
@@ -64,6 +67,28 @@ static int InstallAddon(const std::vector<std::string>& params)
...
+  const std::string& addonid = params[0];
+
+  if (!g_passwordManager.CheckMenuLock(WINDOW_ADDON_BROWSER))
+    return -1;
+
+  AddonPtr addon;
+  if (!CServiceBroker::GetAddonMgr().GetAddon(addonid, addon, ADDON_UNKNOWN, false))
+    return -1;
+
+  auto response = HELPERS::ShowYesNoDialogLines(CVariant{24076}, CVariant{24135}, CVariant{addon->Name()}, CVariant{24136});
```
[the remark this answers is not in the document]
Comment (a1rwulf) [reply xbmc/interfaces/builtins/AddonBuiltins.cpp]: This is a good example to discuss how we will handle this as I still want to help with this untangling one day.
Given the fact that `HELPERS::ShowYesNoDialogLines` is a helper that does `CApplicationMessenger::GetInstance().SendMsg(TMSG_GUI_DIALOG_YESNO, -1, -1, static_cast<void*>(&options))` which is handled by https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIWindowManager.cpp#L1003 
I see to obvious ways to handle this:
* Check for GUI, don't call `ShowYesNoDialogLines` at all
* Implement a stub that handles all those GUI related messages and return YES in there

@FernetMenta What's your opinion on this?

[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/interfaces/builtins/AddonBuiltins.cpp]: 1)
An API call from the outside must not block on user interaction.

2)
We should not allow the outside work to enable addons without ack by user.

A possible solution is that an addons sends an async enable request to Kodi. Kodi handles the request if there is a UI.

File: xbmc/guilib/guiinfo/AddonsGUIInfo.cpp [hunk 5]
```diff
@@ -169,6 +169,14 @@ bool CAddonsGUIInfo::GetBool(bool& value, const CGUIListItem *gitem, int context
       value = CServiceBroker::GetAddonMgr().GetAddon(info.GetData3(), addon) && addon;
       return true;
     }
+    case SYSTEM_ADDON_IS_ENABLED:
+    {
+      value = false;
+      ADDON::AddonPtr addon;
+      if (CServiceBroker::GetAddonMgr().GetAddon(info.GetData3(), addon))
+        value = !CServiceBroker::GetAddonMgr().IsAddonDisabled(info.GetData3());
```
Comment (AlwinEsch) [inline xbmc/guilib/guiinfo/AddonsGUIInfo.cpp]: This "IsAddonDisabled" is by call of GetAddon not direct required. GetAddon give normally without the fourth value only a enabled addon and returns false if not present.

Maybe a more clear call can be:
```cpp
value = (CServiceBroker::GetAddonMgr().IsAddonInstalled(info.GetData3()) &&
        !CServiceBroker::GetAddonMgr().IsAddonDisabled(info.GetData3())
```

*Just a note:* Want to replace this IsAddonDisabled in future to a IsAddonEnabled who is better for me.

**Review by Kp000999 — APPROVED** [review]
#[](url)

File: xbmc/GUIInfoManager.cpp [hunk 6]
```diff
@@ -1183,6 +1183,11 @@ const infomap system_labels[] =  {{ "hasnetwork",       SYSTEM_ETHERNET_LINK_ACT
 ///                  _boolean_,
 ///     Returns true if the specified addon is installed on the system.
 ///   }
+///   \table_row3{   <b>`System.AddonIsEnabled(id)`</b>,
```
Comment (enen92) [inline xbmc/GUIInfoManager.cpp]:
```
///   \table_row3{  <b>`System.AddonIsEnabled(id)`</b>,
///                  \anchor System_AddonIsEnabled
///                  _boolean_,
///     @return **True** if the specified addon is enabled on the system..
///     @param id - The addon Id
///     <p><hr>
///     @skinning_v19 **[New Boolean Condition]** \link System_AddonIsEnabled `System.AddonIsEnabled(id)`\endlink
///     <p>
///   }
```

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 7]
```diff
@@ -441,6 +467,7 @@ CBuiltins::CommandMap CAddonBuiltins::GetOperations() const
            {"addon.default.opensettings", {"Open a settings dialog for the default addon of the given type", 1, OpenDefaultSettings}},
            {"addon.default.set",          {"Open a select dialog to allow choosing the default addon of the given type", 1, SetDefaultAddon}},
            {"addon.opensettings",         {"Open a settings dialog for the addon of the given id", 1, AddonSettings}},
+           {"enableaddon",                {"Enables the specified plugin/script", 1, EnableAddon}},
```
Comment (enen92) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: Missing documentation on the table right above (try to use `@skinning_v19` there too

**Comment by DaveTBlake** [comment]
@ronie it would be nice to get this merged after such a long wait for something wanted by skinners, you are so patient. 

Could you either split out the EnableAddon() builtin to a separate PR as that seems to be the only point of contention, or maybe make the confirmation dialog optional using a parameter to indicate "supress dialog". That at least would allow the caller to avoid the no GUI (TV off etc. ) issue in a way that some of the other builtin command do. But I don't know if others would be satified with that.

**Comment by ronie** [comment]
i'm afraid i have no idea how to optionally suppress a dialog.
can you point me to an example?
do we have some kind of NoGUI condition in Kodi i can check?

**Comment by DaveTBlake** [comment]
Sorry, I was not suggesting anything that clever @ronie, amd probably not sufficient to deal with the issues. 

Updatelibrary uses a parameter in the builtin call
https://github.com/xbmc/xbmc/blob/1842103d59ff7af498244168b57473b6e0133b7e/xbmc/interfaces/builtins/LibraryBuiltins.cpp#L248
and skips the dialogs if the caller set it.

However in this case its would be that a caller could enable an addon without a user confirmation, and maybe that is unacceptable?

**Comment by ronie** [comment]
added the missing documentation as suggested by enen92.

imo this is ready to go in as-is.
if someone has to urge to adopt the code to handle headless Kodi set-ups, feel free to do so.

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 8]
```diff
@@ -22,8 +22,10 @@
 #include "games/tags/GameInfoTag.h"
 #include "guilib/GUIComponent.h"
 #include "guilib/GUIWindowManager.h"
+#include "GUIPassword.h"
```
Comment (Rechi) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: wrong include order
> Included header files have to be sorted (case sensitive) alphabetically [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-headers

Comment (ronie) [reply xbmc/interfaces/builtins/AddonBuiltins.cpp]: thx! missed that one indeed.

**Comment by Rechi** [comment]
@ronie please always set the milestone before or right after merging a PR.

**Pull request merged**